ecosystem/typescript/sdk_v2/src/types/account_address.ts

```diff
@@ -0,0 +1,241 @@
+// dapps and wallets. The enum is meant to make it possible for daps to user their
+// own error messages if they want, but the defaults should still be good, and they're
+// not great right now.
+
+// These options influence fromStr and the isValid* functions. You may use them to make
+// the parsing of those functions strictier. Not setting these options is the best way
+// to comply with AIP 40 but while migrating it may be helpful to be able to enforce
+// stricter behavior. If we agree to make these two enabled by default, I can update
+// the AIP to say that this is what people SHOULD do but they MAY opt in to accepting
+// addresses more leniently in specifically these two ways.
+export type ValidityOptions = {
+  // Only addresses in the long form (64 characters + optionally a leading 0x) will be
+  // considered valid unless the address is special.
+  requireLongFormUnlessSpecial?: boolean;
+
+  // Only addresses in the long form (64 characters + optionally a leading 0x) will be
+  // considered valid, even if the address is special.
+  requireLongForm?: boolean;
+
+  // If set, the address string will be considered invalid if there is no leading 0x.
+  requireLeadingZeroX?: boolean;
+};
+
+// TODO: As part of reviewing this we should discuss these options. I'm partially
+// inclined to say that requireLongFormUnlessSpecial and requireLeadingZeroX should
+// actually be enabled by default. Maayan and Greg to provide their thoughts as well as
+// wallet and dapp folks.
+export const defaultValidityOptions: ValidityOptions = {
+  requireLongFormUnlessSpecial: false,
+  requireLongForm: false,
+  requireLeadingZeroX: false,
+};
+
+export enum AddressInvalidReason {
+  INCORRECT_NUMBER_OF_BYTES = "incorrect_number of bytes",
+  ADDRESS_TOO_SHORT = "address_too_short",
+  ADDRESS_TOO_LONG = "address_too_long",
+  LEADING_ZERO_X_REQUIRED = "leading_zero_x_required",
+  LONG_FORM_REQUIRED = "long_form_required",
+  LONG_FORM_REQUIRED_UNLESS_SPECIAL = "long_form_required_unless_special",
+}
+
+/**
+ * This error is used to explain why an address was invalid.
+ */
+export class AddressInvalidError extends Error {
+  // This provides a programmatic way to access why an address was invalid. Downstream
+  // devs might want to use this to build their own error messages if the default error
+  // messages are not suitable for their use case.
+  public reason: AddressInvalidReason;
+
+  constructor(message: string, reason: AddressInvalidReason) {
+    super(message);
+    this.reason = reason;
+  }
+}
+
+export class AccountAddress {
+  // The number of bytes that make up an account address.
+  static readonly LENGTH: number = 32;
+
+  // The length of an address string in long form without a leading 0x.
+  static readonly LONG_STRING_LENGTH: number = 64;
+
+  // This is the internal representation of an account address.
+  readonly address: Bytes;
+
+  static ADDRESS_ONE: AccountAddress = AccountAddress.fromStr("0x1");
```

Comment:
what are these for?

Reply:
We already had ADDRESS_ONE (under a different name) in the previous AccountAddress class. We tend to deploy all framework code on 0x1, 0x2, 0x3, or 0x4. So it's handy to have constants for those. We do the same thing in Rust and Python.

Reply:
Yes, I use the `0x1` a lot, so much that I have it memorized lol

```typescript
TxnBuilderTypes.AccountAddress.CORE_CODE_ADDRESS
```

But yeah, ADDRESS_ONE probably makes more sense, esp cuz you can extend it to 3 and 4 for the token stuff